tc/core/polyhedral/cuda/memory_promotion_heuristic.cc

```diff
@@ -323,11 +323,138 @@ bool promotionImprovesCoalescing(
   return false;
 }
 
+// Returns a union of all mapping filters to "MappingType" in "scop".
+template <typename MappingType>
+isl::union_set collectMappingsTo(const Scop& scop) {
+  auto root = scop.scheduleRoot();
+  auto domain = scop.domain();
+  auto mappingFilters = detail::ScheduleTree::collect(
+      root, detail::ScheduleTreeType::MappingFilter);
+  mappingFilters = functional::Filter(isMappingTo<MappingType>, mappingFilters);
+  for (auto mf : mappingFilters) {
+    auto filterNode = mf->elemAs<detail::ScheduleTreeElemMappingFilter>();
+    domain = domain.intersect(filterNode->filter_);
+  }
+  return domain;
+}
+
+/*
+ * Check if the band was mapped to threads.
+ * We currently map entire bands to threads and insert a thread-specific marker
+ * immediately after such bands.
+ */
+inline bool isThreadMappedBand(const detail::ScheduleTree* tree) {
+  return matchOne(band(threadSpecific(any())), tree);
+}
+
+/*
+ * Check that only unrolled loops may appear in access subscripts.
+ * Because the scoping point can be above a branching tree, descend into each
+ * leaf of the subtree below the scoping point.  For each leaf, construct an
+ * affine multi-expression containing only those band members between the
+ * scoping point and the leaf that are fully unrolled.  If band members are
+ * mapped to threads, do not take into account the parts that will appear in
+ * subscripts as thread identifiers, i.e. subtract the mapped affine functions
+ * from the schedule affine functions.
+ *
+ * Within each instance of the scope loops, check that loops that are either
+ * unrolled or mapped to threads access a single tensor element in the group
+ * (other loop indices will then not appear in the subscripts, making register
+ * promotion possible).  In other words, check that the relation between the
+ * flat product of prefix, thread-mapped, and unrolled loop indices and
+ * accessed elements is single-valued.
+ * TODO: note that if a group is formed from partially overlapping references,
+ * one must consider per-reference access relation for single-valuedness as
+ * different references may have different values, but all of them remain
+ * independent of non-unrolled loop iterators.
+ */
+bool accessSubscriptsAreUnrolledLoops(
+    const TensorReferenceGroup& group,
+    const detail::ScheduleTree* root,
+    const detail::ScheduleTree* scope,
+    isl::multi_union_pw_aff prefixSchedule,
+    isl::multi_union_pw_aff threadSchedule) {
```

Comment:
You didn't document the meaning of `threadSchedule`, but it seems
to be band schedule from which the thread mapping is derived.
I think you need the actual thread mapping here, i.e., the output of
`MappedScop::threadMappingSchedule`.
Concatenating the prefix schedule, the thread mapping schedule and the unrolled schedule dimensions, should give you the right discriminator.

Reply:
Indeed, this is due to an artifically introduced commit.  The old function was just erased in my original commit rather than going through progressive upates with extra code that just gets deleted by the subsequent commit.

Reply:
On Wed, Jun 13, 2018 at 11:15:26AM -0700, ftynse wrote:
> Indeed, this is due to an artifically introduced commit.  The old function was just eras
ed in my original commit rather than going through progressive upates with extra code that
 just gets deleted by the subsequent commit.

Hmm... from the looks of it, it seems like you did fix this.